plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSource.java

```diff
@@ -326,17 +333,19 @@ public String toString()
             implements ColumnAdaptation
     {
         private final int index;
+        private final UnaryOperator<Block> blockCoercer;
 
-        public SourceColumn(int index)
+        public SourceColumn(int index, UnaryOperator<Block> blockCoercer)
         {
             checkArgument(index >= 0, "index is negative");
             this.index = index;
+            this.blockCoercer = requireNonNull(blockCoercer, "blockCoercer is null");
         }
 
         @Override
         public Block block(Page sourcePage, MaskDeletedRowsFunction maskDeletedRowsFunction, long filePosition, OptionalLong startRowId)
         {
-            Block block = sourcePage.getBlock(index);
+            Block block = blockCoercer.apply(sourcePage.getBlock(index));
```

Comment:
I think this coercion should happen inside the `LazyBlockLoader#load` method on the result of `sourceBlock.getLoadedBlock()` to preserve laziness